### DISCUSSION TITLE
Better support for premade routes through GPX files in route planner/plan a route features.

### DISCUSSION THREAD

**Discussion by adegans** [osmandapp/OsmAnd#12001] [Q&A]
# 🚀 feature request

### Description

When making a route in osmand (iOS) i select where to go and it makes a route in between and that's mostly fine. But is super tedious when the route is longer or has many waypoints, especially if the route is several hundred KM long and has 8+ waypoints.

To make that easier I make routes on my computer with an app "Road Trip Planner" (macOS) which creates GPX files with waypoints, track segments and has an option for POI.

### Describe the solution you'd like

Osmand has a "Plan route" option which can load and save to GPX files, but NO WAY to load existing GPX waypoints as navigation waypoints. Making loading a premade route very very very very tedious and frustrating and basically it doesn't work in it's current state.

The easiest way (I think) is to add an option to the "follow track" option, that if Osmand detects waypoints (Not POI) it'll ask (or have a switch) to use those for a proper route, use the track for the calculated route (somehow) and then use that with the plan a route feature. That would give everyone decent turn-by-turn navigation.
I think Osmand is smart enough to test tracks or segments for a viable route and skip half the calculations (making it fast too).

The actions for the user then could be as easy as; Press Navigation button -> Options -> Follow track -> select GPX file -> Enable switch to use waypoints and track for turn-by-turn navigation (where it shows "pass along whole route") -> press done -> press start. And off the user goes.

### Describe alternatives you've considered

Currently I'm simply following the track, (using the "Follow track" option) but then I don't get turn-by-turn navigation, just osmand dumbly following the thick ugly red line on screen. Which as a basic feature works fine. But turn instructions and road names are not available - Which is what I want solved with this request.

Alternatively I tried importing a GPX with JUST the waypoints and have osmand calculate the in-between routes. But that doesn't work. A 293km route gets 400+km because it keeps looping me around roundabouts and go up and down sidestreets on pretty much every turn while I should be going straight. Or it messes up the waypoint order and sends me all over the place on crossings for no reason. Plus it takes 15+ minutes to calculate... So that's useless in it's current form. However, fixing the routing issue on that end would fix the above request as well, but in a different way...

To test compatibility I've converted the GPX I make/get from the desktop app to a bunch of different formats (no track, convert waypoints to route points, add a track AND route points and a few more combinations) to try and match a route made on osmand. But nothing seems to work as expected. It all comes crashing down when the GPX is ready to go and Osmand doesn't load the waypoints for the route planner and has no easy way to load all waypoints, let alone has a clear understanding of route pins/waypoints and POI pins. Or I have to add waypoints myself over and over again (sometimes they show up as mapmarkers (?) too) or the route calculation goes in illogical and useless ways.

If it matters, I'm using iOS Osmand with the whole country downloaded and I'm currently in Mexico.

Thanks for your consideration!

**Comment by vshcherb** [comment]
If I understood correctly, you have GPX file with waypoints but you want to navigate between them. Indeed the easiest way is to run Plan Route with existing GPX displayed and put points one by one in order you want to visit them.

Actually <rte> - points to navigate through and <wpt> points to be displayed on the map (GPX) and there is no way to easy switch between them.

**Reply by barteis** [reply]
To navigate using imported, generic GPXs, I'm coming around to @vshcherb's method.  First display the imported GPX.  Then use OsmAnd's plan route function to trace the GPX using the appropriate navigation profile (e.g., auto, bicycle, foot).  Generate a second GPX and use  _that_ for navigation.  

@adegans: could you post one of your GPXs here? I'd like to experiment with it.

**Reply by adegans** [reply]
> To navigate using imported, generic GPXs, I'm coming around to @vshcherb's method. First display the imported GPX. Then use OsmAnd's plan route function to trace the GPX using the appropriate navigation profile (e.g., auto, bicycle, foot). Generate a second GPX and use _that_ for navigation.

Right, and what better than to have a button to immediately 'import' or add the GPX waypoints to that route you make in Osmand... As is my request. Osmand doesn't seem to be aware of those points when planning a route at all, making it a stupid and tedious process.

If such a button exists I could (try) to import a GPX with only the Waypoints, no track, and finish the route through Osmand easily. 
Something like : plan route, add waypoints, check the order and move on to calculate the route.

Though, even that step should be unnecessary and creates more hassle than a modern system should do for an already premade route. (in my logic, if Osmand can detect turns and put an arrow on my route for it it can also detect the actual roads and use the track for turn-by-turn instead of having to recalculate it through a 2nd thing.

Example GPX from a few weeks ago - https://disk.stor.fit/index.php/s/DYBMGz2wrcCdfQC

**Reply by barteis** [reply]
I realize this is not the OsmAnd capability you're asking for, but I re-traced your route using the Plan-A-Route feature and generated a new GPX.  Easy to do.  I'm not sure how well it navigates compared to native routing, but it should come close.

https://drive.google.com/file/d/1W-7ayQ4ZMK3TXArqjt29On4JHx8jRAej/view?usp=sharing

**Reply by adegans** [reply]
I'm on a 5500km roadtrip. Having to retrace a new route over an already created route for every section I ride is not an acceptable compromise.
I have perfectly fine routes made on my laptop. And yes, they work fine in Osmand too with the follow track option. But then obviously the turn-by-turn stuff doesn't work. Which is what I'm hoping to fix/get through this topic.

I guess that's too much to ask. So never mind.

**Reply by barteis** [reply]
I'm not a programmer, but this:

> The easiest way (I think) is to add an option to the "follow track" option, that if Osmand detects waypoints (Not POI) it'll ask (or have a switch) to use those for a proper route, use the track for the calculated route (somehow) and then use that with the plan a route feature...I think Osmand is smart enough to test tracks or segments for a viable route and skip half the calculations (making it fast too).

does not sound like a trivial coding problem.

**Reply by barteis** [reply]
I'm not sure why I should be concerned about Garmin compatibility.  I haven't used its products or software in years.

**Comment by adegans** [comment]
Right, on my laptop I make a route - Bunch of waypoints and POI and generate a track for that.
What I'd like to be able to do is use the file a bit more smart/advanced and have Osmand calculate the route based on the supplied track/waypoints.

Importing works fine, but then planning the route is tedious in Osmand. An "add all waypoints to route" button at the very least would improve this alot.
But... Once the waypoints are in, the route calculation between them is wonky for some reason - Sending me over crossings and sidestreets repeatedly sometimes tripling the route distance (300km turns into 900+km) which makes the calculated route useless. If I see it again I'll try to make a screenshot of it.
Having Osmand interpret the track of the GPX as a proper route would be even better - That takes (mostly) away the calculations - It just has to figure out the track and turn it into a route and because it then is a real route we get correct turn by turn navigation with street names and everything.

I think Osmand is smart enough to interpret tracks as routes like this, somehow, but it's just lacking the function/code.

Currently, as you know, importing a GPX with track has Osmand dumbly follow the line and isn't aware of what street you're in. So no street names show up with the direction arrows. While this works fine, it's a bit basic.

I think adding the above features/functions would give Osmand an important edge to compete with Garmin and Tomtom and more importantly make navigation infinitely more easy/better than other mobile apps.

**Comment by vshcherb** [comment]
Hm... OsmAnd is smart enough to interpret tracks with  <rtept> and build routes between these points, you can modify this track using Plan route feature. If you have track with <wpt> they are considered to display only points to pass by.
```
<route>
<rtept>...
<rtept>
</route>
```

**Reply by adegans** [reply]
Except it isn't... AS I explained.

**Comment by vshcherb** [comment]
Converted to a discussion cause it's still not clear exactly how it could be fixed and is it a documentation issue or missing feature.
BTW : please share GPX (as *.gpx.zip here) so it's easier to understand what do you mean and test it.

**Reply by adegans** [reply]
As I wrote several times... 
Adding a "add all waypoints" in the route plan feature would help a lot.
Forgoing the route calculation and interpreting the track from a GPX as an actual route would help a lot too.
Having Osmand do this all automatically as described above would be super nice and most user friendly.

But since you don't even know if this is a documentation thing of a feature request... Nevermind, once again my time and good intentions are wasted on github.
So whatever. I'll just fumble around with the current lacking features, or until someone makes an app that handles GPX content in a more useful way.

I think I've described my idea and wishes quite clearly twice now but if you need more information feel free to ask.

--rant--
It would be nice if you can atleast look at the idea objectively/creatively and see if Osmand could be improved in a meaningful way instead of questioning my request/idea and not undertstanding what we're talking about... Clearly this is a feature request for a missing function. Not a documentation thing.

I'm not demanding you to change Osmand in a big way, I'm merely suggesting to add a thing or two to make navigation better. Simple as that.
--/rant--

**Reply by vshcherb** [reply]
Clearly we have a misunderstanding issue here, but it doesn't make neither me or your right in that situation. Obviously it's impossible to fix something if you don't what's broken. It's a pity that we're not clear enough

**Comment by adegans** [comment]
If your exported GPX has a track, osmand can read it and you can use it using the "Follow track" feature.
Has nothing to do with my request/idea though.

**Comment by Lee-Carre** [comment]
Is this is related to #3908?

**Reply by barteis** [reply]
The capability requested in #3908 is now implemented on the Android version of OsmAnd.  Here's my take:
https://vimeo.com/654264099

**Comment by tvogel** [comment]
I think, the only thing missing is an option to ignore the track and only use waypoints stored in the GPX when loading GPX for routing. OsmAnd already stores all targets as waypoints when saving a route, so that option would get the routing back to the state when saving. Currently it does not do that.

**Reply by ThomasMeister123** [reply]
Yes, I miss this feature too.
Also when you select multiple Waypoints for Navigation in OSMAnd and you will store this. OSMAnd will store the COMPLETE Route (Track), but I will only strore Start, WP1, WPn, Destination.
So I can "load" these (GPX) - File and can calculate the route again (maybe now with bycicle instead of car...)
